### PR TITLE
Fix asset path-traversal outside of roots

### PR DESCRIPTION
**Pull request by shayne**
`/assets/...` requests previously supported path-traversal potentially exposing and serving (private) files outside roots.

**Test plan**

Prior to patching perform the a path-traversal request to the server:
```
GET /assets/../../../../etc/hosts HTTP/1.1
Cache-Control: no-store
Host: 127.0.0.1:8081
Connection: close
Accept-Encoding: gzip
User-Agent: okhttp/2.5.0
```

Apply patch and verify a `404` response with body: `Asset not found`

Test normal asset requests work.

### CODE REVIEW

File: packager/react-packager/src/AssetServer/index.js [hunk 1]
```diff
@@ -129,16 +129,23 @@ class AssetServer {
...
       roots.map(root => {
-        const absPath = path.join(root, dir);
+        // important: we want to resolve root + dir
+        // to ensure the requested path doesn't traverse beyond root
+        const absPath = path.resolve(root, dir);
         return stat(absPath).then(fstat => {
-          return {path: absPath, isDirectory: fstat.isDirectory()};
+          // keep asset requests from traversing files
+          // up from the root (e.g. ../../../etc/hosts)
+          if (absPath.indexOf(root) != 0) {
```
Comment (ide) [inline packager/react-packager/src/AssetServer/index.js]: JS strings now support `startsWith` as part of the standard API

**Comment by ide** [comment]
If this is considered a serious vulnerability, you might also want to look into locking down the filesystem at the OS level: https://www.npmjs.com/package/chroot (probably will have to skip it on Windows, but could at least harden the packager on Linux and OS X).

**Comment by shayne** [comment]
@ide chroot would be great, however I'm thinking multiple roots would cause issues.

**Comment by ide** [comment]
Ah, good point

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by bestander** [comment]
Please note that tests are broken https://circleci.com/gh/facebook/react-native/3736

@ide look for green badges when accepting PR :)

**Comment by shayne** [comment]
Thanks @ide and @bestander – unable to reproduce locally. Will update the diff when I'm able to make progress.

**Comment by mkonicek** [comment]
:+1: Thanks for being on top of this @bestander! We should find some time to make the shipit bot bail with failing tests.

**Comment by ide** [comment]
@bestander oops, sorry about that

**Comment by javache** [comment]
This is also broken on our internal CI. Please fix and land this when possible.

**Comment by shayne** [comment]
Commandeered internal diff. Tests passing. Landing.

**Pull request closed without merging**